### PR TITLE
feat: upgrade ds-card-reader service to edgex 2.2

### PR DESCRIPTION
**Pull request by vli11**
Signed-off-by: Valina Li <valina.li@intel.com>

### CODE REVIEW

File: ds-card-reader/device/physical.go [hunk 1]
```diff
@@ -90,12 +91,18 @@ func (reader *CardReaderPhysical) Status() error {
 func (reader *CardReaderPhysical) Write(commandName string, cardNumber string) {
 	// assemble the values that will be propagated throughout the
 	// device service
+	commandvalue, err := dsModels.NewCommandValueWithOrigin(
+		commandName,
+		edgexcommon.ValueTypeString,
+		cardNumber,
```
Comment (lenny-intel) [inline ds-card-reader/device/physical.go:97]: Maybe resource name `card-data` in the device profile should be `card-number`. Would be a better descriptive name.

File: ds-card-reader/device/physical.go [hunk 2]
```diff
@@ -134,7 +141,7 @@ func (reader *CardReaderPhysical) processDevReadEvents(events []evdev.InputEvent
 			continue
 		}
 
-		reader.Write(common.CommandCardReaderEvent, reader.CardNumber)
+		reader.Write(common.CommandCardData, reader.CardNumber)
```
Comment (lenny-intel) [inline ds-card-reader/device/physical.go:144]:
```suggestion
		reader.Write(common.CommandCardNumber, reader.CardNumber)
```

File: docs_src/automated-checkout-services/device_services.md [hunk 3]
```diff
@@ -29,31 +29,40 @@ The `ds-card-reader` service is much simpler than the `ds-controller-board` serv
...
 ---
 
-#### `PUT`: `http://localhost:48098/api/v1/device/name/ds-card-reader/card-reader-event`
+#### `PUT`: `http://localhost:48098/api/v2/device/name/card-reader/card-data`
 
 The `PUT` API endpoint will push the badge ID (which is sent as part of the API request body) into the card reader device service. Once the card reader device service receives the badge ID, the badge ID will be pushed into the EdgeX bus for other application services to utilize.
 
 ```bash
-curl -X PUT -H "Content-Type: application/json" -d '{"card-reader-event":"0003278200"}' http://localhost:48098/api/v1/device/name/ds-card-reader/card-reader-event
+curl -X PUT -H "Content-Type: application/json" -d '{"card-data":"0003278200"}' http://localhost:48098/api/v2/device/name/card-reader/card-data
```
Comment (lenny-intel) [inline docs_src/automated-checkout-services/device_services.md:37]:
```suggestion
curl -X PUT -H "Content-Type: application/json" -d '{"card-number":"0003278200"}' http://localhost:48098/api/v2/device/name/card-reader/card-number
```

File: ds-card-reader/common/common.go [hunk 4]
```diff
@@ -17,6 +17,6 @@ type CardReaderConfig struct {
 // sent to the card reader driver that performs a health check of the underlying
 // device
 const (
-	CommandCardReaderStatus = "card-reader-status"
-	CommandCardReaderEvent  = "card-reader-event"
+	CommandCardReaderStatus = "status"
+	CommandCardData         = "card-data"
```
Comment (lenny-intel) [inline ds-card-reader/common/common.go:21]:
```suggestion
	CommandCardNumber         = "card-number"
```

File: ds-card-reader/device/virtual_test.go [hunk 5]
```diff
@@ -142,25 +106,22 @@ func TestWrite(t *testing.T) {
...
+	loggingClient := logger.NewMockClient()
 
 	reader := CardReaderVirtual{
 		AsyncCh:       asyncCh,
 		DeviceName:    virtualDeviceName,
 		LoggingClient: loggingClient,
 	}
 
-	reader.Write(common.CommandCardReaderEvent, expectedCardNumberVirtual)
+	reader.Write(common.CommandCardData, expectedCardNumberVirtual)
```
Comment (lenny-intel) [inline ds-card-reader/device/virtual_test.go:124]:
```suggestion
	reader.Write(common.CommandCardNumber, expectedCardNumberVirtual)
```

File: ds-card-reader/driver/driver.go [hunk 6]
```diff
@@ -96,7 +97,7 @@ func (drv *CardReaderDriver) HandleWriteCommands(deviceName string, protocols ma
 	commandName := params[0].DeviceResourceName
 
 	switch commandName {
-	case common.CommandCardReaderEvent:
+	case common.CommandCardData:
```
Comment (lenny-intel) [inline ds-card-reader/driver/driver.go:100]:
```suggestion
	case common.CommandCardNumber:
```

File: ds-card-reader/driver/driver.go [hunk 7]
```diff
@@ -106,7 +107,7 @@ func (drv *CardReaderDriver) HandleWriteCommands(deviceName string, protocols ma
 				return fmt.Errorf(errMsg)
 			}
 
-			drv.CardReader.Write(common.CommandCardReaderEvent, cardNumber)
+			drv.CardReader.Write(common.CommandCardData, cardNumber)
```
Comment (lenny-intel) [inline ds-card-reader/driver/driver.go:110]:
```suggestion
			drv.CardReader.Write(common.CommandCardNumber, cardNumber)
```

File: ds-card-reader/driver/driver_test.go [hunk 8]
```diff
@@ -231,38 +192,27 @@ func TestHandleReadCommands(t *testing.T) {
...
-	assert := assert.New(t)
 	require := require.New(t)
 
 	// prepare some variables for usage in the below tests
 	protocolProperties := map[string]models.ProtocolProperties{}
-	lc := logger.NewClient(cardReaderDeviceServiceName, false, logFile, "DEBUG")
+	lc := logger.NewMockClient()
 
-	successfulCommandVal, err := dsModels.NewCommandValue(common.CommandCardReaderEvent, 0, expectedCardNumber, dsModels.String)
+	successfulCommandVal, err := dsModels.NewCommandValueWithOrigin(common.CommandCardData, edgexcommon.ValueTypeString, expectedCardNumber, 0)
```
Comment (lenny-intel) [inline ds-card-reader/driver/driver_test.go:209]:
```suggestion
	successfulCommandVal, err := dsModels.NewCommandValueWithOrigin(common.CommandCardNumber, edgexcommon.ValueTypeString, expectedCardNumber, 0)
```

File: ds-card-reader/driver/driver_test.go [hunk 9]
```diff
@@ -231,38 +192,27 @@ func TestHandleReadCommands(t *testing.T) {
...
 	require.NoError(err)
 
-	invalidCommandVal, err := dsModels.NewCommandValue(invalid, 0, expectedCardNumber, dsModels.String)
+	invalidCommandVal, err := dsModels.NewCommandValueWithOrigin(invalid, edgexcommon.ValueTypeString, expectedCardNumber, 0)
 	require.NoError(err)
 
-	nonStringCommandVal, err := dsModels.NewCommandValue(common.CommandCardReaderEvent, 0, 0.01, dsModels.ParseValueType("float64"))
+	nonStringCommandVal, err := dsModels.NewCommandValueWithOrigin(common.CommandCardData, edgexcommon.ValueTypeFloat64, 0.01, 0)
```
Comment (lenny-intel) [inline ds-card-reader/driver/driver_test.go:215]:
```suggestion
	nonStringCommandVal, err := dsModels.NewCommandValueWithOrigin(common.CommandCardNumber, edgexcommon.ValueTypeFloat64, 0.01, 0)
```

File: ds-card-reader/driver/driver_test.go [hunk 10]
```diff
@@ -292,8 +242,8 @@ func TestHandleWriteCommands(t *testing.T) {
 			"HandleWriteCommands input param with non-string type",
 			[]*dsModels.CommandValue{nonStringCommandVal},
 			[]dsModels.CommandRequest{{}},
-			[]string{fmt.Sprintf("write command \\\"%v\\\" received non-string value: %v", common.CommandCardReaderEvent, "the data type is not string")},
-			fmt.Errorf("write command \"%v\" received non-string value: %v", common.CommandCardReaderEvent, "the data type is not string"),
+			[]string{fmt.Sprintf("write command \\\"%v\\\" received non-string value: %v", common.CommandCardData, "cannot convert CommandValue of Float64 to String")},
+			fmt.Errorf("write command \"%v\" received non-string value: %v", common.CommandCardData, "cannot convert CommandValue of Float64 to String"),
```
Comment (lenny-intel) [inline ds-card-reader/driver/driver_test.go:246]:
```suggestion
			[]string{fmt.Sprintf("write command \\\"%v\\\" received non-string value: %v", common.CommandCardNumber, "cannot convert CommandValue of Float64 to String")},
			fmt.Errorf("write command \"%v\" received non-string value: %v", common.CommandCardNumbera, "cannot convert CommandValue of Float64 to String"),
```

File: ds-card-reader/res/configuration.toml [hunk 11]
```diff
@@ -1,78 +1,106 @@
...
+CredentialsPath = "credentials"
+
+IncomingTopic = "DataTopic"
+responseTopic = "ResponseTopic"
+
+    [MQTTBrokerInfo.Writable]
+    # ResponseFetchInterval specifies the retry 
+    # interval(milliseconds) to fetch the command response from the MQTT broker
+    ResponseFetchInterval = 500
+
```
Comment (lenny-intel) [inline ds-card-reader/res/configuration.toml:100]: Where did this come from? Now the file we were working with.

File: ds-card-reader/res/profiles/ds-card-reader.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+# Copyright © 2022 Intel Corporation. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+
+name: "rfid-card-reader"
+manufacturer: "Intel"
+description: "For reading a USB rfid card reader"
+
+deviceResources:
+- name: "card-data"
+  description: "Data received from card reader"
```
Comment (lenny-intel) [inline ds-card-reader/res/profiles/ds-card-reader.yaml:12]:
```suggestion
- name: "card-number"
  description: "Card number received from card reader"
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: ds-card-reader/device/physical.go [hunk 13]
```diff
@@ -90,12 +91,18 @@ func (reader *CardReaderPhysical) Status() error {
...
 	// assemble the values that will be propagated throughout the
 	// device service
+	commandvalue, err := dsModels.NewCommandValueWithOrigin(
+		commandName,
+		edgexcommon.ValueTypeString,
+		cardNumber,
+		time.Now().UnixNano()/int64(time.Millisecond),
+	)
+	if err != nil {
+		fmt.Errorf("error on NewCommandValueWithOrigin for %v", commandName)
```
Comment (jim-wang-intel) [inline ds-card-reader/device/physical.go:101]: what are we trying to do with this err formatting?  should we log it here?
Also should we include err details in part of Errorf?

File: ds-card-reader/device/virtual.go [hunk 14]
```diff
@@ -25,12 +26,19 @@ type CardReaderVirtual struct {
...
 	// assemble the values that will be propagated throughout the
 	// device service
+	commandvalue, err := dsModels.NewCommandValueWithOrigin(
+		commandName,
+		common.ValueTypeString,
+		cardNumber,
+		time.Now().UnixNano()/int64(time.Millisecond),
+	)
+	if err != nil {
+		fmt.Errorf("error on NewCommandValueWithOrigin for %v", commandName)
```
Comment (jim-wang-intel) [inline ds-card-reader/device/virtual.go:36]: similar comment above

**Review by jim-wang-intel — CHANGES_REQUESTED** [review]

File: ds-card-reader/device/physical.go [hunk 15]
```diff
@@ -90,12 +91,19 @@ func (reader *CardReaderPhysical) Status() error {
...
 	// assemble the values that will be propagated throughout the
 	// device service
+	commandvalue, err := dsModels.NewCommandValueWithOrigin(
+		commandName,
+		edgexcommon.ValueTypeString,
+		cardNumber,
+		time.Now().UnixNano()/int64(time.Millisecond),
+	)
+	if err != nil {
+		reader.LoggingClient.Error(fmt.Sprintf("error on NewCommandValueWithOrigin for %v", commandName))
```
Comment (jim-wang-intel) [inline ds-card-reader/device/physical.go:101]:
```suggestion
		reader.LoggingClient.Errorf("error on NewCommandValueWithOrigin for %v: %v", commandName, err)
```

File: ds-card-reader/device/virtual.go [hunk 16]
```diff
@@ -25,12 +26,19 @@ type CardReaderVirtual struct {
...
 	// assemble the values that will be propagated throughout the
 	// device service
+	commandvalue, err := dsModels.NewCommandValueWithOrigin(
+		commandName,
+		common.ValueTypeString,
+		cardNumber,
+		time.Now().UnixNano()/int64(time.Millisecond),
+	)
+	if err != nil {
+		reader.LoggingClient.Error(fmt.Sprintf("error on NewCommandValueWithOrigin for %v", commandName))
```
Comment (jim-wang-intel) [inline ds-card-reader/device/virtual.go:36]:
```suggestion
		reader.LoggingClient.Errorf("error on NewCommandValueWithOrigin for %v: %v", commandName, err)
```

**Review by jim-wang-intel — APPROVED** [review]
LGTM

**Review by brian-intel — CHANGES_REQUESTED** [review]
There is a reference to the endpoint in the phase 1 of the documentation that will also need to be updated. that  code can be found here: /docs/phases/phase1.html

**Review by lenny-intel — APPROVED** [review]
LGTM

**Review by brian-intel — APPROVED** [review]

**Pull request merged**